cip/1.accepted/CIP2016-12-14-Constraint-syntax.adoc

```diff
@@ -0,0 +1,285 @@
+
+The constraints define a _domain_ within which the constraint applies.
+The domain is defined by the constraint pattern.
+
+[[uniqueness]]
+=== Node property uniqueness constraint
+
+This constraint enforces that there can not be duplicate values of a certain property for a certain type of node.
+For example, that among nodes labeled with `:Person`, each `email` property must be unique.
+
+==== Syntax
+
+.Grammar definition for node property uniqueness constraint:
+[source, ebnf]
+----
+uniqueness-constraint = "UNIQUE", property-expression, { ",", property-expression } ;
+----
+
+.Example of single-property uniqueness constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT unique_person_email
+FOR (p:Person)
+REQUIRE UNIQUE p.email
+----
+
+.Example of multiple-property uniqueness constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT unique_person_details
+FOR (p:Person)
+REQUIRE UNIQUE p.name, p.email, p.address
+----
+
+==== Semantics
+
+The domain of a property uniqueness constraint is defined as all the nodes with a specific label where the specified property key(s) exist (are non-null).
+When more than one property key is defined as part of the constraint, only nodes where _all_ of the properties exist are part of the domain.
+
+The uniqueness mandates that two distinct nodes within the domain can not have the same combination of values for the defined properties (respectively).
+
+===== Example
+
+Consider the graph created by the following statement:
+
+[source, cypher]
+----
+CREATE (:Color {name: 'white', rgb: 0xffffff})
+CREATE (:Color {name: 'black', rgb: 0x000000})
+CREATE (:Color {name: 'very, very dark grey', rgb: 0x000000}) // rounding error!
+----
+
+Due to the duplication of the `rgb` property, the following attempt at creating a constraint will fail:
+
+[source, cypher]
+----
+CREATE CONSTRAINT only_one_color_per_rgb
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb
+----
+
+Suppose that we would rather like to have one color node per name _and_ RGB value (to work around the rounding errors).
+We could then use the following constraint, without modifying our data:
+
+[source, cypher]
+----
+CREATE CONSTRAINT unique_color_nodes
+FOR (c:Color)
+REQUIRE UNIQUE c.rgb, c.name
+----
+
+[[existence]]
+=== Property existence constraints
+
+Property existence constraints are defined for both nodes and relationships, but the semantics are the same.
+For this reason we will go over both constraints in the same section.
+
+==== Syntax
+
+.Grammar definition for property existence constraint:
+[source, ebnf]
+----
+existence-constraint = "exists", "(", property-expression, ")" ;
+----
+
+.Example of node property existence constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT colors_must_have_rgb
+FOR (c:Color)
+REQUIRE exists(c.rgb)
+----
+
+.Example of relationship property existence constraint:
+[source, cypher]
+----
+CREATE CONSTRAINT rates_have_quality
+FOR ()-[l:RATED]-()
```

Comment:
Are the `()-` parts required? What about 
`CREATE ... FOR [l:RATED]`?
As it is, I keep thinking the directionality and the node labels are relevant, when they're not